filters/DBSCANFilter.cpp

```diff
@@ -0,0 +1,148 @@
+/******************************************************************************
+ * Copyright (c) 2019, Bradley J Chambers (brad.chambers@gmail.com)
+ *
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following
+ * conditions are met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in
+ *       the documentation and/or other materials provided
+ *       with the distribution.
+ *     * Neither the name of Hobu, Inc. or Flaxen Geo Consulting nor the
+ *       names of its contributors may be used to endorse or promote
+ *       products derived from this software without specific prior
+ *       written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+ * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
+ * OF SUCH DAMAGE.
+ ****************************************************************************/
+
+// Adapted from MIT-licensed implemenation provided by
+// https://github.com/intel-isl/Open3D/pull/1038.
+
+#include "DBSCANFilter.hpp"
+
+#include <pdal/KDIndex.hpp>
+
+#include <string>
+#include <unordered_set>
+
+namespace pdal
+{
+
+static StaticPluginInfo const s_info{
+    "filters.dbscan", "DBSCAN Clustering.",
+    "http://pdal.io/stages/filters.dbscan.html"};
+
+CREATE_STATIC_STAGE(DBSCANFilter, s_info)
+
+std::string DBSCANFilter::getName() const
+{
+    return s_info.name;
+}
+
+void DBSCANFilter::addArgs(ProgramArgs& args)
+{
+    args.add("min_points", "Min points per cluster", m_minPoints,
+             static_cast<uint64_t>(6));
+    args.add("eps", "Epsilon", m_eps, 1.0);
+}
+
+void DBSCANFilter::addDimensions(PointLayoutPtr layout)
+{
+    m_cluster =
+        layout->registerOrAssignDim("ClusterID", Dimension::Type::Signed64);
```

Comment:
This type seems huge for this and just adds to memory consumption.

Reply:
In practice, this is huge, and would hardly be useful. That said, poor parameter choice could lead you down the path of landing with one cluster per point. `point_count_t` is uint64, so even larger than what we are providing here, but maybe there is a better way to guard against blowing up, even if the utility of the result is questionable.